### DISCUSSION TITLE
How to spin a x86_64 ubuntu on Apple M1 with lima

### DISCUSSION THREAD

**Discussion by zouchengli** [lima-vm/lima#1797] [Q&A] [answered]
### Description

I want to use lima to run x86 architecture ubuntu system on my arm chip, but I try to run it through rosetta parameters, but it keeps freezing

**Comment by AkihiroSuda** [comment]
Full system mode (robust, but very slow, and may look like freezing)
```
limactl start --vm-type=qemu --arch=x86_64
```

Rosetta mode (less robust, but very fast)
```
limactl start --vm-type=vz --rosetta
lima nerdctl run -it --rm --platform=linux/amd64 ubuntu
```

**Reply by zouchengli** [reply]
[ERROR] RootlessKit failed, see the error messages and https://rootlesscontaine.rs/getting-started/common/

**Reply by zouchengli** [reply]
==> /Users/chengli.zou/.lima/test/serial.log <==
[  119.209749] cloud-init[2214]: [rootlesskit:parent] error: failed to setup UID/GID map: newuidmap 2613 [0 501 1 1 100000 65536] failed: : exec: "newuidmap": executable file not found in $PATH
[  119.218720] cloud-init[2214]: [ERROR] RootlessKit failed, see the error messages and https://rootlesscontaine.rs/getting-started/common/ .
[  119.232813] cloud-init[2214]: LIMA| WARNING: Failed to execute /mnt/lima-cidata/boot/40-install-containerd.sh
[  119.246070] cloud-init[2214]: LIMA| Exiting with code 1
[  119.249074] cloud-init[2214]: 2023-09-15 03:00:35,867 - cc_scripts_per_boot.py[WARNING]: Failed to run module scripts-per-boot (per-boot in /var/lib/cloud/scripts/per-boot)
[  119.252722] cloud-init[2214]: 2023-09-15 03:00:35,870 - util.py[WARNING]: Running module scripts-per-boot (<module 'cloudinit.config.cc_scripts_per_boot' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_scripts_per_boot.py'>) failed

**Reply by AkihiroSuda** [reply, marked as answer]
- In the log, do you see an error related to `apt-get` for installing `newuidmap` (`uidmap` package)?
   If so, is it related to corporate proxies or something similar?
- Does `template://docker-rootful` work?

**Reply by AkihiroSuda** [reply]
Does "Rosetta" mode work? Probably it is what you are looking for.

**Reply by zouchengli** [reply]
yeah unset https_proxy work for me

**Reply by ifaizan** [reply]
Is it possible to create an x86_64 ubuntu with `rosetta` mode instead of the default `qemu` ?

**Comment by zouchengli** [comment]
❯ limactl start --vm-type=qemu --arch=x86_64 --name test
? Creating an instance "test" Proceed with the current configuration
INFO[0001] Attempting to download the image              arch=x86_64 digest="sha256:5ad255d32a30a2cda9f0df19f0a6ce8d6f3c81b63845086a4cb5c43cf97fcb92" location="https://cloud-images.ubuntu.com/releases/23.04/release-20230810/ubuntu-23.04-server-cloudimg-amd64.img"
Downloading the image (ubuntu-23.04-server-cloudimg-amd64.img)
780.00 MiB / 780.00 MiB [-----------------------------------] 100.00% 5.80 MiB/s
INFO[0142] Downloaded the image from "https://cloud-images.ubuntu.com/releases/23.04/release-20230810/ubuntu-23.04-server-cloudimg-amd64.img" 
INFO[0143] Attempting to download the nerdctl archive    arch=x86_64 digest="sha256:3f8c494e3c6a265fe2a3c41ef9d6bc859eeeb22095b3353d3558d8120833a23a" location="https://github.com/containerd/nerdctl/releases/download/v1.5.0/nerdctl-full-1.5.0-linux-amd64.tar.gz"
Downloading the nerdctl archive (nerdctl-full-1.5.0-linux-amd64.tar.gz)
239.66 MiB / 239.66 MiB [-----------------------------------] 100.00% 6.90 MiB/s
INFO[0190] Downloaded the nerdctl archive from "https://github.com/containerd/nerdctl/releases/download/v1.5.0/nerdctl-full-1.5.0-linux-amd64.tar.gz" 
INFO[0202] [hostagent] Starting QEMU (hint: to watch the boot progress, see "/Users/chengli.zou/.lima/test/serial*.log") 
INFO[0202] SSH Local Port: 57436                        
INFO[0202] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0242] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0252] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0262] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0269] [hostagent] The essential requirement 1 of 5 is satisfied 
INFO[0269] [hostagent] Waiting for the essential requirement 2 of 5: "user session is ready for ssh" 
INFO[0297] [hostagent] Waiting for the essential requirement 2 of 5: "user session is ready for ssh" 
INFO[0298] [hostagent] The essential requirement 2 of 5 is satisfied 
INFO[0298] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0338] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0378] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0418] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0458] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0498] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0538] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0579] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0619] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0659] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0699] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0739] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0779] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
FATA[0790] did not receive an event with the "running" status

**Comment by zouchengli** [comment]
**I ran two instances and found that their IP addresses were the same. Why? How do I access this IP through an external host?**

`root@lima-test:~# ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:55:55:3c:5d:38 brd ff:ff:ff:ff:ff:ff
    altname enp0s2
    inet 192.168.5.15/24 metric 100 brd 192.168.5.255 scope global dynamic eth0
       valid_lft 83889sec preferred_lft 83889sec
    inet6 fec0::5055:55ff:fe3c:5d38/64 scope site dynamic mngtmpaddr noprefixroute 
       valid_lft 86171sec preferred_lft 14171sec
    inet6 fe80::5055:55ff:fe3c:5d38/64 scope link 
       valid_lft forever preferred_lft forever

`
`root@lima-test1:~# ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:55:55:b6:bb:47 brd ff:ff:ff:ff:ff:ff
    altname enp0s2
    inet 192.168.5.15/24 metric 100 brd 192.168.5.255 scope global dynamic eth0
       valid_lft 86086sec preferred_lft 86086sec
    inet6 fec0::5055:55ff:feb6:bb47/64 scope site dynamic mngtmpaddr noprefixroute 
       valid_lft 86089sec preferred_lft 14089sec
    inet6 fe80::5055:55ff:feb6:bb47/64 scope link 
       valid_lft forever preferred_lft forever
`

**Reply by jandubois** [reply]
The `eth0` zero interface is purely internal. If you want to reach the VM from the host, then you can specify `--network=vzNAT` and get an additional `lima0` interface that is routable from the host (but not from the external network).

**Reply by zouchengli** [reply]
--network lima:host or --network lima:shared  work for me
But you need to install the network plug-in [socket_vmnet](https://github.com/lima-vm/socket_vmnet)
Also I want to know what is the difference between these two types?

**Reply by jandubois** [reply]
`host` mode only allows communication between the VMs and the host. `shared` mode also allow communication to the external network via this adapter. It doesn't really matter much, as you should always have access to the external network over `eth0`.

This diagram is still pretty accurate; just replace `vde_vmnet` with `socket_vmnet`; they are conceptually the same thing:

![LimaVdeVmnet](https://github.com/lima-vm/lima/assets/78947/1b5ed2c6-699f-4761-9a09-78fbbbcd9933)

@AkihiroSuda Should this image be somewhere in the docs folder in the repo? Or is the link to the KubeCon slides good enough?

**Reply by AkihiroSuda** [reply]
> @AkihiroSuda Should this image be somewhere in the docs folder in the repo? Or is the link to the KubeCon slides good enough?

Yes, but we are switching the docs dir to the website: https://github.com/lima-vm/lima/pull/1754
Also, the image should be eventually rewritten in mermaid.
